File: src/HolzShots/UI/Forms/ExplorerInfoPanel.vb
Imports System.Drawing
Imports System.Windows.Forms

Namespace UI.Forms

    Public Class ExplorerInfoPanel
        Inherits Panel

        Private ReadOnly _upperpen As New Pen(Color.FromArgb(204, 217, 234))
        Private ReadOnly _secondpen As New Pen(Color.FromArgb(217, 227, 240))
        Private ReadOnly _thirdpen As New Pen(Color.FromArgb(232, 238, 247))

        Private ReadOnly _gradientColor1 As Color = Color.FromArgb(237, 242, 249)
        Private ReadOnly _gradientColor2 As Color = Color.FromArgb(241, 245, 251)
        Private ReadOnly _upperBrushPoint As New Point(0, 3)

        Public Overrides Property BackColor As Color = Color.Transparent

        Public Sub New()
            MyBase.New()
            Font = New Font("Segoe UI", 9, FontStyle.Regular, GraphicsUnit.Point)
            ForeColor = Color.FromArgb(255, 30, 57, 91)
        End Sub

        Protected Overrides Sub OnPaint(ByVal e As PaintEventArgs)
            If Enabled Then
                If Width > 0 AndAlso Height > 4 Then
                    Select Case Dock
                        Case DockStyle.Bottom
                            DrawBottom(e.Graphics)
                        Case DockStyle.Top
                            DrawTop(e.Graphics)
                        Case DockStyle.Fill
                            DrawFill(e.Graphics)
                        Case DockStyle.Left
                            DrawLeft(e.Graphics)
                        Case DockStyle.Right
                            DrawRight(e.Graphics)
                        Case DockStyle.None
                            DrawFill(e.Graphics)
                    End Select
                End If
            Else
                e.Graphics.FillRectangle(New SolidBrush(Color.FromArgb(240, 240, 240)), DisplayRectangle)
            End If
            MyBase.OnPaint(e)
        End Sub

#Region "DrawStuff"

        Private Function CreateBrushTopDock() As Brush
            Return New Drawing2D.LinearGradientBrush(_upperBrushPoint, New Point(0, MyBase.Height - 3), _gradientColor2, _gradientColor1)
        End Function
        Private Function CreateBrushBottomDock() As Brush
            Return New Drawing2D.LinearGradientBrush(_upperBrushPoint, New Point(0, MyBase.Height - 3), _gradientColor1, _gradientColor2)
        End Function
        Private Function CreateBrushFillDock() As Brush
            Return New Drawing2D.LinearGradientBrush(_upperBrushPoint, New Point(MyBase.Width, MyBase.Height), _gradientColor1, _gradientColor2)
        End Function
        Private Function CreateBrushLeftDock() As Brush
            Return New Drawing2D.LinearGradientBrush(_upperBrushPoint, New Point(MyBase.Width, MyBase.Height), _gradientColor1, _gradientColor2)
        End Function
        Private Function CreateBrushRightDock() As Brush
            Return New Drawing2D.LinearGradientBrush(_upperBrushPoint, New Point(MyBase.Width, MyBase.Height), _gradientColor2, _gradientColor1)
        End Function

        Private Sub DrawBottom(ByVal g As Graphics)
            With g
                If MyBase.Height > 6 Then
                    .FillRectangle(CreateBrushTopDock, 0, 3, MyBase.Width, MyBase.Height - 3)
                End If
                .DrawLine(_upperpen, 0, 0, MyBase.Width, 0)
                .DrawLine(_secondpen, 0, 1, MyBase.Width, 1)
                .DrawLine(_thirdpen, 0, 2, MyBase.Width, 2)
            End With
        End Sub
        Private Sub DrawTop(ByVal g As Graphics)
            With g
                .FillRectangle(CreateBrushBottomDock, 0, 0, MyBase.Width, MyBase.Height - 2)
                .DrawLine(_thirdpen, 0, MyBase.Height - 3, MyBase.Width, MyBase.Height - 3)
                .DrawLine(_secondpen, 0, MyBase.Height - 2, MyBase.Width, MyBase.Height - 2)
                .DrawLine(_upperpen, 0, MyBase.Height - 1, MyBase.Width, MyBase.Height - 1)
            End With
        End Sub
        Private Sub DrawFill(ByVal g As Graphics)
            With g
                .FillRectangle(CreateBrushFillDock, 0, 0, MyBase.Width, MyBase.Height)
                .DrawRectangle(_upperpen, 0, 0, MyBase.Width - 1, MyBase.Height - 1)
                .DrawRectangle(_secondpen, 1, 1, MyBase.Width - 3, MyBase.Height - 3)
                .DrawRectangle(_thirdpen, 2, 2, MyBase.Width - 5, MyBase.Height - 5)
            End With
        End Sub
        Private Sub DrawLeft(ByVal g As Graphics)
            With g
                .FillRectangle(CreateBrushLeftDock, 0, 0, MyBase.Width - 3, MyBase.Height - 1)
                .DrawLine(_upperpen, MyBase.Width - 1, 0, MyBase.Width - 1, MyBase.Height - 1)
                .DrawLine(_secondpen, MyBase.Width - 2, 0, MyBase.Width - 2, MyBase.Height - 1)
                .DrawLine(_thirdpen, MyBase.Width - 3, 0, MyBase.Width - 3, MyBase.Height - 1)
            End With
        End Sub
        Private Sub DrawRight(ByVal g As Graphics)
            With g
                .FillRectangle(CreateBrushRightDock, 0, 0, MyBase.Width, MyBase.Height)
                .DrawLine(_upperpen, 0, 0, 0, MyBase.Height)
                .DrawLine(_secondpen, 1, 0, 1, MyBase.Height)
                .DrawLine(_thirdpen, 2, 0, 2, MyBase.Height)
            End With
        End Sub

#End Region
    End Class
End Namespace

File: src/HolzShots.Windows/Forms/ExplorerInfoPanel.cs
using System.Drawing;
using System.Drawing.Drawing2D;
using System.Windows.Forms;

namespace HolzShots.Windows.Forms;

public class ExplorerInfoPanel : Panel
{
    private static readonly Pen _upperPen = new(Color.FromArgb(204, 217, 234));
    private static readonly Pen _secondPen = new(Color.FromArgb(217, 227, 240));
    private static readonly Pen _thirdPen = new(Color.FromArgb(232, 238, 247));

    private static readonly Color _gradientColor1 = Color.FromArgb(237, 242, 249);
    private static readonly Color _gradientColor2 = Color.FromArgb(241, 245, 251);
    private static readonly Point _upperBrushPoint = new(0, 3);

    public override Color BackColor { get; set; } = Color.Transparent;

    public ExplorerInfoPanel()
    {
        Font = new Font("Segoe UI", 9, FontStyle.Regular, GraphicsUnit.Point);
        ForeColor = Color.FromArgb(255, 30, 57, 91);
    }

    protected override void OnPaint(PaintEventArgs e)
    {
        ArgumentNullException.ThrowIfNull(e);

        base.OnPaint(e);

        if (!Enabled)
        {
            e.Graphics.FillRectangle(new SolidBrush(Color.FromArgb(240, 240, 240)), DisplayRectangle);
            return;
        }

        if (Width <= 0 || Height <= 4)
            return;

        switch (Dock)
        {
            case DockStyle.Bottom:
                DrawBottom(e.Graphics);
                break;
            case DockStyle.Top:
                DrawTop(e.Graphics);
                break;
            case DockStyle.Fill:
                DrawFill(e.Graphics);
                break;
            case DockStyle.Left:
                DrawLeft(e.Graphics);
                break;
            case DockStyle.Right:
                DrawRight(e.Graphics);
                break;
            case DockStyle.None:
                DrawFill(e.Graphics);
                break;
            default:
                DrawFill(e.Graphics);
                break;
        }
    }


    private Brush CreateBrushTopDock() => new LinearGradientBrush(_upperBrushPoint, new Point(0, Height - 3), _gradientColor2, _gradientColor1);
    private Brush CreateBrushBottomDock() => new LinearGradientBrush(_upperBrushPoint, new Point(0, Height - 3), _gradientColor1, _gradientColor2);
    private Brush CreateBrushFillDock() => new LinearGradientBrush(_upperBrushPoint, new Point(Width, Height), _gradientColor1, _gradientColor2);
    private Brush CreateBrushLeftDock() => new LinearGradientBrush(_upperBrushPoint, new Point(Width, Height), _gradientColor1, _gradientColor2);
    private Brush CreateBrushRightDock() => new LinearGradientBrush(_upperBrushPoint, new Point(Width, Height), _gradientColor2, _gradientColor1);

    private void DrawBottom(Graphics g)
    {
        if (Height > 6)
            g.FillRectangle(CreateBrushTopDock(), 0, 3, Width, Height - 3);
        g.DrawLine(_upperPen, 0, 0, Width, 0);
        g.DrawLine(_secondPen, 0, 1, Width, 1);
        g.DrawLine(_thirdPen, 0, 2, Width, 2);
    }

    private void DrawTop(Graphics g)
    {
        g.FillRectangle(CreateBrushBottomDock(), 0, 0, Width, Height - 2);
        g.DrawLine(_thirdPen, 0, Height - 3, Width, Height - 3);
        g.DrawLine(_secondPen, 0, Height - 2, Width, Height - 2);
        g.DrawLine(_upperPen, 0, Height - 1, Width, Height - 1);
    }
    private void DrawFill(Graphics g)
    {
        g.FillRectangle(CreateBrushFillDock(), 0, 0, Width, Height);
        g.DrawRectangle(_upperPen, 0, 0, Width - 1, Height - 1);
        g.DrawRectangle(_secondPen, 1, 1, Width - 3, Height - 3);
        g.DrawRectangle(_thirdPen, 2, 2, Width - 5, Height - 5);
    }
    private void DrawLeft(Graphics g)
    {
        g.FillRectangle(CreateBrushLeftDock(), 0, 0, Width - 3, Height - 1);
        g.DrawLine(_upperPen, Width - 1, 0, Width - 1, Height - 1);
        g.DrawLine(_secondPen, Width - 2, 0, Width - 2, Height - 1);
        g.DrawLine(_thirdPen, Width - 3, 0, Width - 3, Height - 1);
    }
    private void DrawRight(Graphics g)
    {
        g.FillRectangle(CreateBrushRightDock(), 0, 0, Width, Height);
        g.DrawLine(_upperPen, 0, 0, 0, Height);
        g.DrawLine(_secondPen, 1, 0, 1, Height);
        g.DrawLine(_thirdPen, 2, 0, 2, Height);
    }
}
